File: src/accessiweather/forecast_confidence.py
"""
Forecast confidence indicator from cross-source agreement.

Compares key metrics (temperature, precipitation probability) across
multiple forecast sources and produces a human-readable confidence level.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum

from accessiweather.models.weather import SourceData

__all__ = [
    "ForecastConfidenceLevel",
    "ForecastConfidence",
    "calculate_forecast_confidence",
]

# Thresholds for HIGH / MEDIUM confidence
_TEMP_HIGH = 5.0  # °F — both temp and precip within this → HIGH
_TEMP_MED = 10.0  # °F — only temp considered → MEDIUM
_PRECIP_HIGH = 15.0  # % — precip spread for HIGH
_PRECIP_MED = 25.0  # % — precip spread for MEDIUM

# Human-readable display names for source identifiers
_SOURCE_DISPLAY_NAMES: dict[str, str] = {
    "nws": "NWS",
    "openmeteo": "Open-Meteo",
    "pirateweather": "Pirate Weather",
}


def _display_name(source_id: str) -> str:
    """Return a human-readable name for a source identifier."""
    return _SOURCE_DISPLAY_NAMES.get(source_id.lower(), source_id.title())


def _join_sources(names: list[str]) -> str:
    """Join source names naturally: 'A and B' or 'A, B, and C'."""
    if len(names) == 0:
        return ""
    if len(names) == 1:
        return names[0]
    if len(names) == 2:
        return f"{names[0]} and {names[1]}"
    return ", ".join(names[:-1]) + f", and {names[-1]}"


class ForecastConfidenceLevel(Enum):
    """Three-tier confidence classification."""

    HIGH = "High"
    MEDIUM = "Medium"
    LOW = "Low"


@dataclass
class ForecastConfidence:
    """Result of a cross-source confidence calculation."""

    level: ForecastConfidenceLevel
    rationale: str
    sources_compared: int
    source_names: list[str] = field(default_factory=list)


def _valid_sources(sources: list[SourceData]) -> list[SourceData]:
    """Return only sources that have usable forecast data."""
    return [s for s in sources if s.success and s.forecast is not None and s.forecast.has_data()]


def calculate_forecast_confidence(sources: list[SourceData]) -> ForecastConfidence:
    """
    Compute a confidence level by comparing the first forecast period across sources.

    Args:
        sources: List of :class:`~accessiweather.models.weather.SourceData` objects
            (one per weather provider).

    Returns:
        A :class:`ForecastConfidence` describing the level, rationale, and how
        many valid sources were included in the calculation.

    """
    valid = _valid_sources(sources)
    n = len(valid)
    names = [_display_name(s.source) for s in valid]
    joined = _join_sources(names)

    if n == 0:
        return ForecastConfidence(
            level=ForecastConfidenceLevel.LOW,
            rationale="No forecast sources available",
            sources_compared=0,
            source_names=[],
        )

    if n == 1:
        return ForecastConfidence(
            level=ForecastConfidenceLevel.MEDIUM,
            rationale=f"Based on {joined} — single source, no cross-reference available",
            sources_compared=1,
            source_names=names,
        )

    # --- 2+ sources: extract first-period metrics ---
    temps: list[float] = []
    precips: list[float] = []

    for s in valid:
        assert s.forecast is not None  # already filtered above
        if s.forecast.periods:
            p0 = s.forecast.periods[0]
            if p0.temperature is not None:
                temps.append(p0.temperature)
            if p0.precipitation_probability is not None:
                precips.append(p0.precipitation_probability)

    temp_spread = (max(temps) - min(temps)) if len(temps) >= 2 else 0.0
    precip_spread = (max(precips) - min(precips)) if len(precips) >= 2 else None

    # --- Apply thresholds ---
    if precip_spread is not None:
        # Both temperature AND precipitation data available
        if temp_spread <= _TEMP_HIGH and precip_spread <= _PRECIP_HIGH:
            return ForecastConfidence(
                level=ForecastConfidenceLevel.HIGH,
                rationale=f"{joined} agree on temperature and precipitation",
                sources_compared=n,
                source_names=names,
            )
        if temp_spread <= _TEMP_MED or precip_spread <= _PRECIP_MED:
            return ForecastConfidence(
                level=ForecastConfidenceLevel.MEDIUM,
                rationale=f"Moderate agreement between {joined}",
                sources_compared=n,
                source_names=names,
            )
        return ForecastConfidence(
            level=ForecastConfidenceLevel.LOW,
            rationale=f"{joined} show significant disagreement on temperature or precipitation",
            sources_compared=n,
            source_names=names,
        )

    # Temperature-only comparison
    if temp_spread <= _TEMP_HIGH:
        return ForecastConfidence(
            level=ForecastConfidenceLevel.HIGH,
            rationale=f"{joined} agree on temperature",
            sources_compared=n,
            source_names=names,
        )
    if temp_spread <= _TEMP_MED:
        return ForecastConfidence(
            level=ForecastConfidenceLevel.MEDIUM,
            rationale=f"Moderate temperature agreement between {joined}",
            sources_compared=n,
            source_names=names,
        )
    return ForecastConfidence(
        level=ForecastConfidenceLevel.LOW,
        rationale=f"{joined} show significant temperature disagreement",
        sources_compared=n,
        source_names=names,
    )

File: rust/crates/aw-core/src/forecast_confidence.rs
//! Forecast confidence from cross-source agreement.
//! Port of `accessiweather/forecast_confidence.py`; the result types live in
//! [`crate::model`].

use crate::model::{
    Forecast, ForecastConfidence, ForecastConfidenceLevel, ForecastPeriod, SourceData,
};

const TEMP_HIGH: f64 = 5.0;
const TEMP_MED: f64 = 10.0;
const PRECIP_HIGH: f64 = 15.0;
const PRECIP_MED: f64 = 25.0;

/// Human-readable name for a source id (`str.title()` for unknown ids).
pub fn source_display_name(source_id: &str) -> String {
    match source_id.to_lowercase().as_str() {
        "nws" => "NWS".into(),
        "openmeteo" => "Open-Meteo".into(),
        "pirateweather" => "Pirate Weather".into(),
        _ => title_case(source_id),
    }
}

/// Python's `str.title()`: upper-case letters that follow a non-letter.
fn title_case(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut previous_cased = false;
    for c in s.chars() {
        if previous_cased {
            out.extend(c.to_lowercase());
        } else {
            out.extend(c.to_uppercase());
        }
        previous_cased = c.is_alphabetic();
    }
    out
}

/// "A", "A and B", "A, B, and C".
fn join_sources(names: &[String]) -> String {
    match names {
        [] => String::new(),
        [one] => one.clone(),
        [a, b] => format!("{a} and {b}"),
        [rest @ .., last] => format!("{}, and {last}", rest.join(", ")),
    }
}

/// First period with a temperature that is not a night period, else the
/// first period (sources disagree on whether periods[0] is a high or a low).
fn representative_period(forecast: &Forecast) -> Option<&ForecastPeriod> {
    forecast
        .periods
        .iter()
        .find(|p| p.temperature.is_some() && !p.name.to_lowercase().contains("night"))
        .or(forecast.periods.first())
}

fn spread(values: &[f64]) -> Option<f64> {
    if values.len() < 2 {
        return None;
    }
    let max = values.iter().copied().fold(f64::NEG_INFINITY, f64::max);
    let min = values.iter().copied().fold(f64::INFINITY, f64::min);
    Some(max - min)
}

/// `calculate_forecast_confidence`: compare the representative period's
/// temperature and precipitation chance across the sources that have a forecast.
pub fn calculate_forecast_confidence(sources: &[SourceData]) -> ForecastConfidence {
    let valid: Vec<&SourceData> = sources
        .iter()
        .filter(|s| s.success && s.forecast.as_ref().is_some_and(Forecast::has_data))
        .collect();
    let names: Vec<String> = valid
        .iter()
        .map(|s| source_display_name(&s.source))
        .collect();
    let joined = join_sources(&names);
    let n = valid.len() as i64;
    let result = |level, rationale: String| ForecastConfidence {
        level,
        rationale,
        sources_compared: n,
        source_names: names.clone(),
    };

    match n {
        0 => {
            return result(
                ForecastConfidenceLevel::Low,
                "No forecast sources available".into(),
            )
        }
        1 => {
            return result(
                ForecastConfidenceLevel::Medium,
                format!("Based on {joined} — single source, no cross-reference available"),
            )
        }
        _ => {}
    }

    let mut temps = Vec::new();
    let mut precips = Vec::new();
    for s in &valid {
        if let Some(period) = s.forecast.as_ref().and_then(representative_period) {
            temps.extend(period.temperature);
            precips.extend(period.precipitation_probability);
        }
    }
    let temp_spread = spread(&temps).unwrap_or(0.0);

    if let Some(precip_spread) = spread(&precips) {
        if temp_spread <= TEMP_HIGH && precip_spread <= PRECIP_HIGH {
            return result(
                ForecastConfidenceLevel::High,
                format!("{joined} agree on temperature and precipitation"),
            );
        }
        if temp_spread <= TEMP_MED || precip_spread <= PRECIP_MED {
            return result(
                ForecastConfidenceLevel::Medium,
                format!("Moderate agreement between {joined}"),
            );
        }
        return result(
            ForecastConfidenceLevel::Low,
            format!("{joined} show significant disagreement on temperature or precipitation"),
        );
    }

    if temp_spread <= TEMP_HIGH {
        result(
            ForecastConfidenceLevel::High,
            format!("{joined} agree on temperature"),
        )
    } else if temp_spread <= TEMP_MED {
        result(
            ForecastConfidenceLevel::Medium,
            format!("Moderate temperature agreement between {joined}"),
        )
    } else {
        result(
            ForecastConfidenceLevel::Low,
            format!("{joined} show significant temperature disagreement"),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::golden::{self, field};

    #[test]
    fn golden_forecast_confidence() {
        let cases = golden::load("confidence/cases.json");
        for case in cases.as_array().unwrap() {
            let sources: Vec<SourceData> = field(case, "sources");
            let expected: ForecastConfidence = field(case, "confidence");
            assert_eq!(
                calculate_forecast_confidence(&sources),
                expected,
                "{}",
                case["name"]
            );
        }
    }

    #[test]
    fn display_names_and_joining() {
        assert_eq!(source_display_name("OpenMeteo"), "Open-Meteo");
        assert_eq!(source_display_name("visual_crossing"), "Visual_Crossing");
        assert_eq!(source_display_name("abc123def"), "Abc123Def");
        let names: Vec<String> = ["A", "B", "C"].iter().map(|s| s.to_string()).collect();
        assert_eq!(join_sources(&names), "A, B, and C");
        assert_eq!(join_sources(&names[..2]), "A and B");
    }
}
